### PR TITLE
deepcopy: make slice+map generation uniform and nestable

### PR DESCRIPTION
**Pull request by sttts**
This PR uses `g.generateFor` inside slice and map deepcopy generation instead of special casing of some non-nested types. `g.generateFor` is supposed to be nestable such that we can generate code for arbitrarily nested types like  `[]*map[string]***[]*string`.

Moreover, this PR
- adds reflection deepcopy unit tests of the output
- adds a benchmark
- makes the generator fail early and hard when there is an unknown type case, instead of best effort code generation with `out = in.DeepCopy()`.

Fixes https://github.com/kubernetes/gengo/issues/109.

### CODE REVIEW

**Comment by sttts** [comment]
Due to the slightly more complex generated map code, this deserved a benchmark.

**Comment by lavalamp** [comment]
The expanded test case looks awesome. This LGTM if you squash

**Comment by sttts** [comment]
@lavalamp will add a benchmark and give @thockin a chance to look at the updated output per commit. Will squash then at the end.

**Comment by sttts** [comment]
And I switched to hard errors instead of best effort `out = in.DeepCopy()` generation. Want to test-vendor into kube to see whether we missed cases.

File: examples/deepcopy-gen/output_tests/maps/zz_generated.go [hunk 1]
```diff
@@ -104,13 +104,152 @@ func (in *Ttest) DeepCopyInto(out *Ttest) {
...
 		for key, val := range *in {
+			var outVal *string
 			if val == nil {
 				(*out)[key] = nil
 			} else {
-				x := new(string)
-				(*out)[key] = x
-				*x = *val
+				in, out := &val, &outVal
+				if *in == nil {
```
Comment (thockin) [inline examples/deepcopy-gen/output_tests/maps/zz_generated.go]: This is redundant with L107/108 above

**Comment by sttts** [comment]
Added deepcopy tests with fuzzing, comparing against reflection based deepcopy.

And the benchmarks, before the structural changes:

```
300000	      4655 ns/op
3000000	       588 ns/op
20000000	       107 ns/op
```

and after

```
300000	      4509 ns/op
3000000	       584 ns/op
20000000	       106 ns/op
```

File: examples/deepcopy-gen/output_tests/maps/zz_generated.go (hunk 1 above)
Comment (sttts) [reply examples/deepcopy-gen/output_tests/maps/zz_generated.go]: fixed. Pointers had a redundant nil check.

**Comment by sttts** [comment]
@lavalamp @thockin squashed. ptal.

**Pull request merged**